Fix memory leak in sandbox extension code

pvollan opened the pull request:
#### d2e94b018a5a96215b37f681db435b0f99b5e9a1
<pre>
Fix memory leak in sandbox extension code
<a href="https://bugs.webkit.org/show_bug.cgi?id=288172">https://bugs.webkit.org/show_bug.cgi?id=288172</a>
<a href="https://rdar.apple.com/145238621">rdar://145238621</a>

Reviewed by Chris Dumez.

The function SandboxExtensionImpl::sandboxExtensionForType currently returns a C string that needs to
be deallocated by the caller. Not all callers of this function are deallocating the memory. This patch
addresses this issue by changing the function to return a CString. The caller does no longer have to
deallocate the C string returned from the sandbox functions that issued the extension. The deallocation
of the C string is now done inside the method itself.

* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::sandboxExtensionForType):
* Source/WebKit/Shared/SandboxExtension.h:

Canonical link: <a href="https://commits.webkit.org/290834@main">https://commits.webkit.org/290834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f739e59f8c209dcdd656de4bd4b5c198f9f8b137

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91149 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10692 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96151 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41907 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11098 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19010 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70051 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27573 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94150 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8458 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82586 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50377 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8229 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41045 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78549 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/175 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98134 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18351 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13456 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79060 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18610 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78417 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78263 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22768 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/117 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11543 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14415 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18353 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23670 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18077 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19853 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
```diff
@@ -81,8 +81,10 @@
-char* SandboxExtensionImpl::sandboxExtensionForType(const char* path, SandboxExtension::Type type, std::optional<audit_token_t> auditToken, OptionSet<SandboxExtension::Flags> flags)
+CString SandboxExtensionImpl::sandboxExtensionForType(const char* path, SandboxExtension::Type type, std::optional<audit_token_t> auditToken, OptionSet<SandboxExtension::Flags> flags)
 {
+    char* sandboxExtension = nullptr;
```
I think we could use this:
```
auto sandboxExtension = [&] {
    switch (type) {
    case SandboxExtension::Type::ReadOnly:
        return std::unique_ptr<char, decltype(free)*>(sandbox_extension_issue_file(APP_SANDBOX_READ, path, extensionFlags), free);
    // ...
    }
})();
```

pvollan replied:
Ah, very nice, will fix.

Thanks for reviewing!

cdumez commented on Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
```diff
@@ -91,27 +93,39 @@
+
+    CString sandboxExtensionString(sandboxExtension);
+    if (sandboxExtension)
+        free(sandboxExtension);
```
Then we wouldn't need this.

cdumez commented on Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
```diff
@@ -91,27 +93,39 @@
+        break;
     }
+
+    CString sandboxExtensionString(sandboxExtension);
```
`sandboxExtension.get()` with my proposal above.

cdumez commented on Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
```diff
@@ -81,37 +81,42 @@
-            return nullptr;
+CString SandboxExtensionImpl::sandboxExtensionForType(const char* path, SandboxExtension::Type type, std::optional<audit_token_t> auditToken, OptionSet<SandboxExtension::Flags> flags)
+{
+    auto sandboxExtensionString = [&] {
```
The `auto*` here is a `char*` that needs freeing. This is too error prone, we want a smart pointer *right away*, like I originally suggested.

pvollan replied:
Ah, yes, good point, fixed in latest patch.

Thanks for reviewing!

cdumez reviewed (approved):
LGTM

pvollan committed:
Fix memory leak in sandbox extension code
https://bugs.webkit.org/show_bug.cgi?id=288172
rdar://145238621

Reviewed by Chris Dumez.

The function SandboxExtensionImpl::sandboxExtensionForType currently returns a C string that needs to
be deallocated by the caller. Not all callers of this function are deallocating the memory. This patch
addresses this issue by changing the function to return a CString. The caller does no longer have to
deallocate the C string returned from the sandbox functions that issued the extension. The deallocation
of the C string is now done inside the method itself.

* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::sandboxExtensionForType):
* Source/WebKit/Shared/SandboxExtension.h:

Canonical link: https://commits.webkit.org/290834@main
